editor/components/higher-order/if-post-type-supports/index.js

```diff
@@ -0,0 +1,57 @@
+/**
+ * External dependencies
+ */
+import { connect } from 'react-redux';
+import { castArray, get, some } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { ifPropsVerify, withAPIData } from '@wordpress/components';
+import { compose, getWrapperDisplayName } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { getCurrentPostType } from '../../../store/selectors';
+
+export function checkIfPostTypeSupports( postType, supportKeys ) {
+	return !! some(
+		castArray( supportKeys ), key => get( postType, [ 'data', 'supports', key ], false )
+	);
+}
+
+const applyConnect = connect( ( state ) => {
+	return {
+		postTypeSlug: getCurrentPostType( state ),
+	};
+} );
+
+const applyWithAPIData = withAPIData( ( { postTypeSlug } ) => {
+	return {
+		postType: postTypeSlug ? `/wp/v2/types/${ postTypeSlug }?context=edit` : undefined,
+	};
+} );
+
+const applyIfPropsVerify = ( supportKeys ) => ifPropsVerify( ( { postType } ) => checkIfPostTypeSupports( postType, supportKeys ) );
+
+/**
+ * A Higher Order Component that provides a way to render a component or not
+ * depending on post type supports of the current post being edited.
+ *
+ * @param   {string|Array} supportKeys An array or string. The component is rendered
+ *                                     if post type supports at least one of the support keys passed.
+ * @returns {Component}                Wrapped component.
+ */
+export default function ifPostTypeSupports( supportKeys ) {
+	return ( OriginalComponent ) => {
+		const ComponentWithApply = compose(
+			applyConnect,
+			applyWithAPIData,
+			applyIfPropsVerify( supportKeys ),
+		)( OriginalComponent );
+		const WrappedComponent = ( props ) => <ComponentWithApply { ...props } />;
+		WrappedComponent.displayName = getWrapperDisplayName( WrappedComponent, 'ifPostTypeSupports' );
+		return WrappedComponent;
+	};
+}
```

Comment:
I think I 'd prefer a renderProp where you pass the result of the check because we have use-case where we want to do `if else`.

```js
<IfPostTypeSupports
  feature="comments" 
  renderIfTrue={ () => <div>Awesome</div> }
  renderIfFalse={ () => <div>Oh No!</div> }
/>
```

I think the higher order component could still be exposed as a special case though.

Reply:
Hm, I agree this would be difficult to implement with higher-order components.

The prop naming doesn't jive with me, and I'd be more inclined toward a function-as-child alternative, but for no particularly strong reasons:

```jsx
<IfPostTypeSupports feature="comments">
	{ ( isSupported ) => (
		<div>{ isSupported ? 'Awesome' : 'Oh No!' }</div>
	) }
</IfPostTypeSupports>
```

_(Though in this trivial example it does help for reducing code duplication on shared wrappers, although in the real world this would probably be wrapping the `<IfPostTypeSupports />` anyways)_

Reply:
Hi @youknowriad, @aduth, I thought a little and I think we can keep the simplicity of HoC and still allow rendering an else condition.
If we use the HoC like:
ifPostTypeSupports( 'thumbnail' ),

The HoC does not render the component, used in most of the cases.

But if we use the HoC like:
ifPostTypeSupports( 'thumbnail', 'isThumbnailSupported' ),
The component is rendered with the additional prop isThumbnailSupported passed so the user of the HoC can implement its if/else logic.

Let me know your thoughts on this approach.

Reply:
Yeah, that works pretty well. Would it be the standard usage? Or would there be a default inferred usage that if the second argument isn't specified, the component doesn't render at all if the condition isn't met?†

_† And if so, is this potentially confusing in its inconsistency?_

Reply:
My idea was that the components don't render at all if the second argument isn't specified. It may be a not easy to guess behavior but if we document it I think users of the component will get familiar with it.